Publicize: Add REST API Take 3

mdawaffe opened the pull request:
Built on top of #10600.

Obsoletes #10487, #10398.

@tyxla has a PR for the client side that makes the Gutenberg extension work with this PR: Automattic/wp-calypso#28463

#### Changes proposed in this Pull Request:

Adds REST API endpoints for Publicize data:
* Services
* Connections
* Connection Test Results
* A Post's Connections

#### Testing instructions:

##### Automated Tests
1. `phpunit --group rest-api`
2. `phpunit --group publicize`

##### Manual Tests

The best way to test is to make REST API requests to a test site. The simplest way to do that is to install either the [REST API Console](https://wordpress.org/plugins/rest-api-console/) plugin or the [Basic Authentication](https://github.com/WP-API/Basic-Auth) plugin.

###### As an Admin:

1. In the Classic (non-Gutenberg editor) create a post, but do not publish it.
2. Add a custom Publicize message. Save as Draft.
3. Make an authenticated REST API request: `GET /wp-json/wpcom/v2/publicize/services`. You should see Facebook, Twitter, etc.
4. Make an authenticated REST API request: `GET /wp-json/wpcom/v2/publicize/connections`. You should see the Publicize Connections you have forged for your site+user.
5. Make an authenticated REST API request: `GET /wp-json/wpcom/v2/publicize/connection-test-results`.  You should see something very similar to the previous request but with additional properties detailing the results of the connection test for each connection.
6. Make an authenticated REST API request: `GET /wp-json/wp/v2/posts/{the-post-you-drafted}/`. You should see a new field: `jetpack_publicize_connections` detailing the connection settings for that post. You should also see the custom Publicize Message in a new `meta` field: `jetpack_publicize_message`.

###### As a Contributor:
1. In the Classic (non-Gutenberg editor) create a post, but do not publish it.
2. Make an authenticated REST API request: `GET /wp-json/wpcom/v2/publicize/services`. You should see an HTTP 403.
3. Make an authenticated REST API request: `GET /wp-json/wpcom/v2/publicize/connections`. You should see an HTTP 403.
4. Make an authenticated REST API request: `GET /wp-json/wpcom/v2/publicize/connection-test-results`.  You should see an HTTP 403.
5. Make an authenticated REST API request: `GET /wp-json/wp/v2/posts/{the-post-you-drafted}/`. You should see a new field: `jetpack_publicize_connections` **that is an empty array**, and a new `meta` field: `jetpack_publicize_message` **that is an empty string**.

Needs testing back to WordPress 4.8.

#### Proposed changelog entry for your changes:

Provide API for accessing Publicize data.

ockham committed:
Add get_publicize_conns_test_results() function

ockham committed:
Add JSDoc to get_publicize_conns_test_results

mdawaffe committed:
Typo

mdawaffe committed:
Publicize: Helper function for permissions check

We have a filter so that other plugins can change the capability required
to view/edit Publicize data.

Rather than having to remember to call that filter all the time, let's
provide a post-specific helper method.

mdawaffe committed:
Publicize: Explicitly register `publicize` support for posts

Rather than making posts an exception, this makes some post_type stuff
easier in the API.

mdawaffe committed:
Publicize API: Add Publicize Message as Post Meta in API

mdawaffe committed:
Publicize API: Post's Connections

mdawaffe committed:
Publicize API: Connections, Services Endpoints

Separate Connections and Connections Test Results data into separate
endpoints.

mdawaffe committed:
Publicize API: Documentation Typos

mdawaffe committed:
Publicize API: Whitespace fixes

ockham commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+	 * @return WP_REST_Response suitable for 1-page collection
+	 */
+	public function get_items( $request ) {
+		global $publicize;
```
How about setting a `$publicize` instance var upon construction? We'll probably still use the global for that, but at least we'd move responsibility to the consumer, arguably providing _somewhat_ better encapsulation, and possibly facilitating cleaner separation of concerns in a hypothetical future without that global...

mdawaffe replied:
The `$publicize` global isn't set by the time this class is constructed (at least, the way these files are currently loaded). Getting things loaded in the right order would take a fair bit of work. Let's wait for that hypothetical refactor in the future ;)

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+			'type' => 'object',
+			'properties' => $this->get_connection_schema_properties() + array(
+				'test_success' => array(
+					'description' => __( 'Did the Publicize Connection test pass?', 'jetpack' ),
```
It makes sense that we capitalize `Publicize`, but I think we should make an audit in the whole PR and lowercase any `Connection` or `Service` that occurs in the middle of a phrase.

mdawaffe replied:
I wanted to make it clear it was a Publicize Connection, etc. and not an HTTP connection, Jetpack connection, etc. I'm happy to change it, though.

tyxla replied:
Thanks, given the context I'm good with it as-is 👍

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	/**
+	 * Adds the test results properties to the Connection schema.
```
We could add a `@return` here.

mdawaffe replied:
Thanks. Fixed in ea8627f70.

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connection-test-results.php:
```diff
@@ -0,0 +1,111 @@
+			$test_result = $test_results_by_unique_id[$item['id']];
+
+			foreach ( $mapping as $field => $test_result_field ) {
+				$item[$field] = $test_result[$test_result_field];
```
Seems like we have some missing spaces here. Perhaps we could run phpcs here?

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+ *
+ * [
+ *   { # Connnection Object. See schema for more detail.
+ *     id:           (string)  Connection unique_id
```
Question: is there a good reason why `unique_id` is a string?

mdawaffe replied:
Just that I don't trust Jetpack and WordPress.com both to always return an integer. It's supposed to be an opaque value anyway, so using a string gives us some future flexibility.

Also, if we ever pass 4294967295 connections, an integer would break some clients :) (I'm sure there are other fields where this is much more likely to happen :)).

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+	 * Helper for generating schema. Used by this endpoint and by the
+	 * Connection Test Result endpoint.
+	 *
+	 * @internal
```
Some of these methods could use `@return`

mdawaffe replied:
ea8627f70

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+				$connection_data = $connection_meta['connection_data'];
+
+				$items[] = array(
+					'id' => $publicize->get_connection_unique_id( $connection ),
```
Do we need to cast to string here to ensure that the ID will be string?

mdawaffe replied:
Since we've defined the schema, I *think* it's handled by core, but I agree we should do it here anyway.

In theory, `Publicize::get_connection_unique_id()` should be returning a string anyway, but… I suspect it's not. We should probably add casting upstream.

Fixed here in 500e82d87.

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+					'id' => $publicize->get_connection_unique_id( $connection ),
+					'service_name' => $service_name,
+					'display_name' => $publicize->get_display_name( $service_name, $connection ),
+					'global' => 0 == $connection_data['user_id'],
```
Nit: would be nice to align those `=>`s vertically.

mdawaffe replied:
37fe2b442

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+					'id' => $publicize->get_connection_unique_id( $connection ),
+					'service_name' => $service_name,
+					'display_name' => $publicize->get_display_name( $service_name, $connection ),
+					'global' => 0 == $connection_data['user_id'],
```
If we want to stick with the non-strict equality check, let's add an inline comment that explains why it's there.

mdawaffe replied:
847e409ff

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+	 * @param WP_REST_Request $request
+	 * @return array filtered $connection
+	 */
+	function prepare_item_for_response( $connection, $request ) {
```
We seem to be lacking a specific visibility modifier here. It'll default to `public`, but since this looks internal, we might want to encapsulate it.

mdawaffe replied:
It's public: https://core.trac.wordpress.org/browser/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-controller.php?rev=43571#L186

Fixed in f717306c1.

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,163 @@
+		$response_data = array();
+		foreach ( $connection as $field => $value ) {
+			if ( in_array( $field, $fields, true ) ) {
+				$response_data[$field] = $value;
```
We're missing some spaces inside the brackets here.

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-services.php:
```diff
@@ -0,0 +1,140 @@
+			'type' => 'object',
+			'properties' => array(
+				'name' => array(
+					'description' => __( 'Alphanumeric identifier for the Publicize Service', 'jetpack' ),
```
Is this really an identifier? Looks like it is the name, or "slug" of the service.

mdawaffe replied:
Yeah - it's a slug. I don't think anyone outside of WordPress developers uses that term, though, do they?

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-services.php:
```diff
@@ -0,0 +1,140 @@
+			'type' => 'object',
+			'properties' => array(
+				'name' => array(
+					'description' => __( 'Alphanumeric identifier for the Publicize Service', 'jetpack' ),
```
A language nit: "for the Publicize Service" looks ambiguous - it's not clear whether this means "for Publicize, itself", or "for a service that Publicize supports".

mdawaffe replied:
Yeah - I can see the confusion. Since this is the "publicize/services" endpoint, I think it's OK, though.

Happy to change it if you have a suggestion, of course :)

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-services.php:
```diff
@@ -0,0 +1,140 @@
+		 * We also need add_submenu_page(), as the URLs for connecting each service
+		 * rely on the `sharing` menu subpage being present.
+		 */
+		include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
```
I'd love it if we could get rid of any assumptions that this code is running in wp-admin context. Not in this PR, of course.

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-services.php:
```diff
@@ -0,0 +1,140 @@
+		}
+
+		$response = rest_ensure_response( $services );
+		$response->header( 'X-WP-Total', count( $services ) );
```
We seem to be adding those headers several times - do we want to have a helper method do it for us?

mdawaffe replied:
Let's follow the core `WP_REST_Controller` pattern and be explicit.

https://core.trac.wordpress.org/browser/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php?rev=43584#L340

tyxla commented on _inc/lib/core-api/wpcom-endpoints/publicize-services.php:
```diff
@@ -0,0 +1,140 @@
+		$response_data = array();
+		foreach ( $service as $field => $value ) {
+			if ( in_array( $field, $fields, true ) ) {
+				$response_data[$field] = $value;
```
Missing spaces inside the brackets.

tyxla commented on _inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php:
```diff
@@ -0,0 +1,256 @@
+ * @since 6.8.0
+ */
+class WPCOM_REST_API_V2_Post_Publicize_Connections_Field extends WPCOM_REST_API_V2_Field_Controller {
+	protected $object_type = 'post';
```
This will always be overriden by the `register_fields()` call. Do we need it for anything before that?

mdawaffe replied:
No - I left it from an earlier draft since I thought it made it more clear what the property was for.

tyxla commented on _inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php:
```diff
@@ -0,0 +1,256 @@
+			$output_connection = array();
+			foreach ( $properties as $property  ) {
+				if ( isset( $connection[$property] ) ) {
+					$output_connection[$property] = $connection[$property];
```
We are missing spaces inside the brackets.

tyxla reviewed (commented):
This looks great, and tests well with https://github.com/Automattic/wp-calypso/pull/28463 (which is already merged FWIW)! 💯 

I've left some minor comments, and will refrain from approving before the PR is finished, but generally, this feels very close to go. Nice work!

mdawaffe committed:
Publicize API: PHPDoc fixes

mdawaffe commented:
Rebased.

mdawaffe committed:
Publicize API: Cast `id` to string

Just to be sure.

mdawaffe committed:
Publicize API: Code style

mdawaffe committed:
Publicize API: Exlpain loose equality check

mdawaffe committed:
Pubicize API: method visibility

mdawaffe commented:
Rebased again.

mdawaffe committed:
Publicize API: Linter fixes

mdawaffe commented:
Rebased again again.

mdawaffe commented:
Rebased.

mdawaffe committed:
Publicize API: Don't fatal when `WP_REST_Controller::get_fields_for_response()` is not defined.

It was first available in WordPress 4.9.6.

mdawaffe commented:
D20681-code largely merged on WordPress.com. Only the new post field is left to merge.

mdawaffe commented:
Let's get this merged now-ish. @Automattic/poseidon Merge at will, and you can be in charge of any remaining bugs in future PRs. Feel free to commandeer D20681-code as well.

(Hopefully TravisCI passes :))

tyxla reviewed (dismissed):
LGTM 🚢

ockham commented:
Hmm unit tests are failing...

ockham commented on _inc/lib/core-api/wpcom-endpoints/publicize-connections.php:
```diff
@@ -0,0 +1,178 @@
+	 */
+	protected function get_connection_schema_properties() {
+		return array(
+			'id'           => array(
```
Whitespace seems off

ockham replied:
Oh, aligned with arrows further down. Nevermind then.

mdawaffe committed:
Publicize API: Clean up test post types

They seem to be breaking the `WP_Test_Jetpack_Sync_Functions` tests
but only in Travis...?

mdawaffe commented:
e48f2f287 is an attempt at fixing the unit tests.

lezama reviewed (approved):
Great work 🙇

jeherve commented:
Cherry-picked to `branch-6.8` in 5a99a5fee0fa206dd2a0a987299a592545d03d86
